validation/validate.py

```diff
@@ -0,0 +1,215 @@
+            print(e)
+    return d
+
+
+def type_matches(value: Any, expected_type: str) -> bool:
+    '''
+    Checks whether value has the type expected_type.
+    If yes, returns True, otherwise False.
+
+    Allowed types T are: 'str', 'int', 'float', 'complex', 'bool',
+                         as well as 'List[T]'.
+
+    For any other types the function raises ValueError.
+    '''
+    allowed_basic_types = ['str', 'int', 'float', 'complex', 'bool']
+    allowed_list_types = ['List[{}]'.format(t) for t in allowed_basic_types]
+    allowed_types = allowed_basic_types + allowed_list_types
+
+    # first verify whether expected_type is allowed
+    if expected_type not in allowed_types:
+        raise ValueError('could not recognize expected_type: {}'.format(expected_type))
+
+    expected_type_is_list = re.match(r"^List\[\w+\]$", expected_type) is not None
+
+    if expected_type_is_list:
+        # make sure that value is actually a list
+        if not type(value).__name__ == 'list':
+            return False
+
+        # iterate over each element of the list
+        # and check whether it matches T
+        list_element_type = re.search(r"^List\[(\w+)\]$", expected_type).group(1)
+        for element in value:
+            if not type(element).__name__ == list_element_type:
+                return False
+
+        return True
+    else:
+        # expected type is a basic type
+        # this is the simpler case
+        return type(value).__name__ == expected_type
+
+
+def test_type_matches():
+
+    # str
+    assert type_matches('a string', 'str')
+    assert not type_matches(1.0, 'str')
+
+    # float and int
+    assert type_matches(1.0, 'float')
+    assert type_matches(1.0e-8, 'float')
+    assert type_matches(1, 'int')
+    assert not type_matches(1, 'float')
+
+    # complex
+    assert type_matches(1.0+1.0j, 'complex')
+
+    # bool
+    assert type_matches(True, 'bool')
+    assert type_matches(False, 'bool')
+    assert not type_matches(0, 'bool')
+
+    # lists
+    assert type_matches([1, 2, 3], 'List[int]')
+    assert not type_matches((1, 2, 3), 'List[int]')
+    assert not type_matches([1, 2, 3], 'List[float]')
+    assert type_matches([1.0, 2.0, 3.0], 'List[float]')
+    assert not type_matches([1.0, 2.0, 3], 'List[float]')
+
+    # unexpected type input
+    with pytest.raises(ValueError):
+        assert type_matches('example', 'weird')
+        assert type_matches('example', 'List[strange]')
+
+
+class InputError(Exception):
+    pass
+
+
+class TemplateError(Exception):
+    pass
+
+
+def validate_node(input_dict, template_dict):
+    '''
+    Verify a node. A node is either the root of the input or it is a section.
+    The node can contain keywords or other sections which we traverse recursively.
+
+    This function modifies input_dict and returns it.
+
+    Can raise either InputError or TemplateError.
+    '''
+    input_sections = [x for x in input_dict if type(input_dict[x]).__name__ == 'dict']
```

Comment:
I would use `isinstance` rather than `type(x).__name__ ==`

Reply:
I agree - I will adjust.

Reply:
Same as above - resolving this but I keep the above conversation.